vnext/ReactUWP/Utils/LocalBundleReader.cpp

```diff
@@ -22,9 +23,23 @@ std::future<std::string> LocalBundleReader::LoadBundleAsync(const std::string& b
   co_await winrt::resume_background();
 
   auto file = co_await winrt::Windows::Storage::StorageFile::GetFileFromApplicationUriAsync(uri);
-  auto hdata = co_await winrt::Windows::Storage::FileIO::ReadTextAsync(file);
 
-  co_return facebook::react::unicode::utf16ToUtf8(std::wstring(hdata.data()));
+  // Read the buffer manually to avoid a Utf8 -> Utf16 -> Utf8 encoding roundtrip.
+  auto fileBuffer { co_await winrt::Windows::Storage::FileIO::ReadBufferAsync(file) };
+  auto dataReader{ winrt::Windows::Storage::Streams::DataReader::FromBuffer(fileBuffer) };
+  std::vector<uint8_t> data(fileBuffer.Length() + 1);
+
+  // manually place the null byte at the end.
+  data[fileBuffer.Length()] = 0;
+
+  // Construct the array_view to slice into the first fileBuffer.Length bytes.
+  // DataReader.ReadBytes will read as many bytes as are present in the array_view.
+  // The backing vector has fileBuffer.Length() + 1 bytes, without an explicit end it will read 1 byte to many and throw.
+  winrt::array_view<uint8_t> arrayView{ data.data(), data.data() + fileBuffer.Length() };
+  dataReader.ReadBytes(arrayView);
+  dataReader.Close();
+
+  co_return std::string(reinterpret_cast<const char*>(data.data()));
```

Comment:
can also use the 2nd arg, data.size() in this std::string() ctor so it doesn't have to scan the buffer to find the null terminator.

Did you have to create the array_view to limit how much it was reading?  everywhere else we use readbytes we just pass the vector to it and internally it makes the array_view.  maybe a difference between buffer.length and unconsumedbufferlength?

Reply:
I will update to give the size of the string.

yes, the array_view was specifically to limit how much would be read. 
I didn't see an overload for [ReadBytes ](https://docs.microsoft.com/en-us/uwp/api/windows.storage.streams.datareader.readbytes) allowing me to specify how many bytes to read, and this was the first way I could come up with that would prevent reading past the end of the buffer. Open to any suggestions though.

The crux of the issue is just that I needed to have fileLength + 1 space to leave room for the null byte. 

I could try pre-allocating the string, then give an array_view of the string to read bytes. that would also address the string copying you had mentioned.
But i'm not sure if/how to avoid using the array_view entirely.

Reply:
Updated to pre-allocate the string and avoid the copy. This shaves off another handful of ms from the previous versions.

Maybe you already stated this and I misunderstood. But do you know if there is any mechanism we can use to skip the initial fill of the string?
In the trace you can see the initial memset and moving the file into the string.
![image](https://user-images.githubusercontent.com/48095233/59527423-660be080-8e90-11e9-86ab-6ff20526c076.png)

Reply:
I assume no to maintain security and not reveal the previous contents of that memory. But I figured it's worth asking

Reply:
I don't think that's avoidable using the STL wrappers.  If you drop to just having a buffer of memory instead of vector its possible to avoid, but more error prone.  I think there are more utf16->utf8->utf16 conversions as we call into Chakra, unavoidable with the current chakra API + jsi api

Reply:
yeah there is another one (that i know of) where we need to go from utf8 to utf16 to execute the js. but this one seemed entirely avoidable so I wanted to take care of it

Reply:
any more changes you'd like to see?